### PR TITLE
feat: add copy-if-not-exists to azure

### PR DESCRIPTION
**Pull request by roeap**
This PR implements `copy-if-not.exists` to azure store.

This PR replaces #26 

@tustvold - as you pointed out, much simpler to implement with blob ... :)

### CODE REVIEW

File: src/azure.rs [hunk 1]
```diff
@@ -393,12 +392,49 @@ impl ObjectStore for MicrosoftAzure {
...
+                }
+            })?;
         Ok(())
     }
+}
 
-    async fn copy_if_not_exists(&self, _from: &Path, _to: &Path) -> Result<()> {
-        Err(crate::Error::NotImplemented)
+impl MicrosoftAzure {
+    fn get_from_url(&self, from: &Path) -> Result<reqwest::Url> {
```
Comment (tustvold) [inline src/azure.rs:430]:
```suggestion
    fn get_copy_from_url(&self, from: &Path) -> Result<reqwest::Url> {
```

And perhaps a docstring, I initially thought this was a method for converting from a URL, and was then confused :laughing:

File: src/lib.rs [hunk 2]
```diff
@@ -103,6 +103,18 @@ pub trait ObjectStore: std::fmt::Display + Send + Sync + Debug + 'static {
...
+
+    /// Move an object from one path to another in the same object store.
+    ///
+    /// Will return an error if the destination already has an object.
+    async fn rename_if_not_exists(&self, from: &Path, to: &Path) -> Result<()> {
+        match self.copy_if_not_exists(from, to).await {
+            Ok(_) => self.delete(from).await,
+            // TODO do we need to do more rigorous error checking?
+            other => other,
+        }?;
```
Comment (tustvold) [inline src/lib.rs:115]:
```suggestion
        self.copy_if_not_exists(from, to).await?;
        self.delete(from).await
```
I don't think you need any additional error checking that isn't already performed by copy_if_not_exists?

**Review by tustvold — COMMENTED** [review]
Looking good, left some minor comments, CI also appears to be failing

**Comment by roeap** [comment]
Played around with the emulator a bit more and ended up opening Azure/Azurite#1543 to see if I am missing something or its in fact an error.

**Comment by tustvold** [comment]
I suspect that might be the case. Perhaps you could disable the failing assertion for the Azurite store, there should be an example of this already for range requests iirc.

**Comment by roeap** [comment]
made the relevant assert conditional and - maybe not surprisingly 😆 - the CI is passing now..

**Review by tustvold — APPROVED** [review]
Looks good to me 😀

**Pull request merged**